addon/lib/StudyTelemetryCollector.jsm

```diff
@@ -21,15 +22,266 @@ const BASERESOURCE = "esper-pioneer-shield-study";
+      TelemetryEnvironment.onInitialized(),
+      TelemetrySession.delayedInit(),
+    ]).then(() => {
+
+      // Wait for the gather-telemetry notification before we gather telemetry
+      const gatherPromise = PromiseUtils.defer();
+      Services.obs.addObserver(gatherPromise.resolve, "gather-telemetry");
+      TelemetrySession.observe(null, "idle-daily", null);
+      gatherPromise.promise.then(() => {
+
+        console.log("Received gather-telemetry notification.");
+        Services.obs.removeObserver(gatherPromise.resolve, "gather-telemetry");
+
+        try {
+
+          this.collectAndSendTelemetry();
+
+        } catch (ex) {
+          // TODO: how are errors during study execution reported?
+          // this.studyUtils.telemetryError();
+          Components.utils.reportError(ex);
+        }
+
+
+      });
+
+    });
+
+  }
+
+  async collectAndSendTelemetry() {
+
+    const telemetryPingEnvelopeBasedAttributes = StudyTelemetryCollector.collectTelemetryPingEnvelopeBasedAttributes();
+    const telemetrySubsessionPayloadBasedAttributes = StudyTelemetryCollector.collectTelemetrySubsessionPayloadBasedAttributes();
+
+    console.log("telemetryPingEnvelopeBasedAttributes", telemetryPingEnvelopeBasedAttributes);
+    console.log("telemetrySubsessionPayloadBasedAttributes", telemetrySubsessionPayloadBasedAttributes);
+
+    StudyTelemetryCollector.collectPlacesDbBasedAttributes().then((placesDbBasedAttributes) => {
+
+      console.log("placesDbBasedAttributes", placesDbBasedAttributes);
+
+      const shieldPingAttributes = {
+        event: "telemetry-payload",
+        ...telemetryPingEnvelopeBasedAttributes,
+        ...telemetrySubsessionPayloadBasedAttributes,
+        ...placesDbBasedAttributes,
+      };
+
+      const shieldPingPayload = StudyTelemetryCollector.createShieldPingPayload(shieldPingAttributes);
+
+      console.log("shieldPingPayload", shieldPingPayload);
+
+      this.studyUtils.telemetry(shieldPingPayload);
+
+    });
+
+  }
+
+  // TODO: @glind: move to shield study utils?
+  static createShieldPingPayload(shieldPingAttributes) {
+
+    const shieldPingPayload = {};
+
+    // shield ping attributes must be strings
+    for (const attribute in shieldPingAttributes) {
+      let attributeValue = shieldPingAttributes[attribute];
+      if (typeof attributeValue === "undefined") {
+        attributeValue = "null";
+      }
+      if (typeof attributeValue === "object") {
+        attributeValue = JSON.stringify(attributeValue);
+      }
+      if (typeof attributeValue !== "string") {
+        attributeValue = String(attributeValue);
+      }
+      shieldPingPayload[attribute] = attributeValue;
+    }
+
+    return shieldPingPayload;
+
+  }
+
+  /**
+   * These attributes are already sent as part of the telemetry ping envelope
+   * @returns {{}}
+   */
+  static collectTelemetryPingEnvelopeBasedAttributes() {
+
+    const telemetrySessionCurrentPingData = TelemetryController.getCurrentPingData();
```

Comment:
Calling `getCurrentPingData` can be slow and should be avoided unless you need subsession data and measurements (e.g. scalars, histograms and measurements in general). Since you're only using the environment below, you could as well just get it using `TelemetryEnvironment.currentEnvironment`.

Reply:
Thanks, refactored